### PR TITLE
Salesforce: Support AuraDefinitionBundle & LightningComponentBundle SALTO-852

### PR DESCRIPTION
**Pull request by omrilit**
The `targetConfigs` field of LightningComponentBundle is not typed in SF and arrives in base64Binary encoding.
[LightningComponentBundle Metadata API definition](https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_lightningcomponentbundle.htm)
[LightningComponentBundle targetConfigs dev guide definition](https://developer.salesforce.com/docs/component-library/documentation/en/lwc/lwc.reference_configuration_tags)
There are basically 3 options to represent this field:
1. Create its type definition manually, so it will look like all other fields.
 **_This PR suggestion_**
2. Use the `unknown` type for it so it will look like all other fields but won't be tight to a fragile type definition and won't fail on validations.
**_Eventually this one is not chosen due to:
-In case we receive an array argument with only a single string, in the NaCL it'll appear as a string since it's not typed.
-In order to omit the prefix of the XML attribute in the NaCLs, we should be familiar with its fields def, so there is no reason to hard code them without defining it as the type_**
3. Set the retrieved XML part of the `targetConfigs` content in a different static file and wrap it as a string.
**_That might look weird to the user as in SFDX all of the metadata XML is united and not separated. On the other hand, this way we won't have the '\_ATTRIBUTE\_' prefix flag for  the `targets` inner value (we use it in order to send it back upon deploy as an attribute)._**

LightningComponentBundle instance NaCL example (option 2):
image
*we can modify the '\_ATTRIBUTE\_'  prefix to something else if you have any suggestions.

LightningComponentBundle type NaCL example (option 2):
image

AuraDefinitionBundle instance NaCL example:
image

@ori-moisis @eladmallel @tomermevorach @gilhoffer  thoughts?

### CODE REVIEW

**Comment by gilhoffer** [comment]
I'm okay with the approach, a few comments though:
1. `_ATTRIBUTE_TARGETS` should be a list of strings, not a single string.
2. Not sure I understand why can't we have _'_ATTRIBUTE_TARGETS` just named `targets` ? (and have a specific filter sending it an as attribute?)

**Comment by omrilit** [comment]
> _ATTRIBUTE_TARGETS should be a list of strings, not a single string.

Since `targetConfigs` isn't typed it's written just as received from the retrieve result. In this case only a single value was returned for the `targets` attribute and thus it appears as a string. In case more will be returned it will be written as an array.

> Not sure I understand why can't we have _'_ATTRIBUTE_TARGETS` just named `targets` ? (and have a specific filter sending it an as attribute?)

We can have a filter that handles these attributes manually and takes care to remove and add the prefix both upon fetch & deploy. There are 14 possible attributes as documented in [LightningComponentBundle targetConfigs dev guide definition
](https://developer.salesforce.com/docs/component-library/documentation/en/lwc/lwc.reference_configuration_tags) We should take into consideration that in case this type will be changed these attributes still might be presented as above, but I do agree that it worths a specific filter handling.

**Comment by gilhoffer** [comment]
Yep, there are 14 attributes (across the different tags), but as it is well documented, not sure how fragile it is really. I would try and prettify it, at least for the attributes we are aware of (I'm okay with a generic handling for tags which are not documented / we are not aware of).

**Comment by ori-moisis** [comment]
I thought we talked about implementing this in the transformer, not in a filter, exactly for this reason.

also, I would strongly advise creating a separate PR for adding the "unknown" built in type to core. it should not be mixed in with another feature, especially not with an adapter feature

**Comment by omrilit** [comment]
That's why I added it in a different commit that is not about to be squashed :)
I can move it to a different PR, NP.

**Comment by ori-moisis** [comment]
I'm not a huge fan of the multiple feature PR approach with multiple commits.
It is ok when the features are tiny and somewhat related, but adding a new builtin type is not a small thing that should be hidden away in a PR with a different subject, it is a significant change to the language.

I think the same goes for adding a "preDeploy" hook on filters, feels like it would justify its own PR if we choose to go that route (though I feel less strongly about that one given it is an internal implementation change in the adapter and has less external effect of things)

**Comment by ori-moisis** [comment]
re:
> Not sure I understand why can't we have _'_ATTRIBUTE_TARGETSjust namedtargets` ? (and have a specific filter sending it an as attribute?)

Doing so would require us to have a complete type for this field, and since It seems like salesforce does not describe the type of this field though any API, we'd have to hard code this type manually.
I think we should try to avoid defining the type manually mostly because I believe it leaves too much room for error (the documentation doesn't seem too comprehensive to me), that is why I am in favor of adding the `unknown` type and trying to do as little as possible in the transformation code.
for me the lack of confidence in that documentation outweighs the ugliness of the nacls (though I would suggest a shorter, less ALL CAPS prefix, like `attr_` to mitigate that a bit), @gilhoffer wdyt?

File: packages/salesforce-adapter/src/constants.ts [hunk 1]
```diff
@@ -25,6 +25,7 @@ export const ADMIN_PROFILE = 'Admin'
 export const NAMESPACE_SEPARATOR = '__'
 export const API_NAME_SEPARATOR = '.'
 export const CUSTOM_OBJECT_ID_FIELD = 'Id'
+export const XML_ATTRIBUTE_PREFIX = '_ATTRIBUTE_'
```
[the remark this answers is not in the document]
Comment (omrilit) [reply packages/salesforce-adapter/src/constants.ts:28]: Agreed f2f to leave it there as the `transformPrimitives` should be aware of it as well.

File: packages/salesforce-adapter/src/transformers/transformer.ts [hunk 2]
```diff
@@ -929,6 +926,14 @@ const convertXsdTypeFuncMap: Record<string, ConvertXsdTypeFunc> = {
   'xsd:long': Number,
 }
 
+// Salesforce returns nulls in metadata API as objects like { $: { 'xsi:nil': 'true' } }
+// and in retrieve API like <activateRSS xsi:nil="true"/>
+// which is transformed to { `${XML_ATTRIBUTE_PREFIX}xsi:nil`): 'true' }
+const isXsiNil = (value: Value): boolean =>
+  _.isObject(value)
+    && (_.get(value, ['$', 'xsi:nil']) === 'true'
+      || _.get(value, `${XML_ATTRIBUTE_PREFIX}xsi:nil`) === 'true')
```
[the remark this answers is not in the document]
Comment (omrilit) [reply packages/salesforce-adapter/src/transformers/transformer.ts:935]: Agreed f2f to leave it there

File: packages/salesforce-adapter/src/transformers/xml_transformer.ts [hunk 3]
```diff
@@ -173,6 +183,137 @@ export const toRetrieveRequest = (files: ReadonlyArray<FileProperties>): Retriev
 
 export type MetadataValues = MetadataInfo & Values
 
+const addContentFieldAsStaticFile = (values: Values, valuePath: string[], content: Buffer,
+  fileName: string): void => {
+  _.set(values, valuePath, content.toString() === HIDDEN_CONTENT_VALUE
+    ? content.toString()
+    : new StaticFile({
+      filepath: fileName.replace(PACKAGE, SALESFORCE),
```
[the remark this answers is not in the document]
Comment (omrilit) [reply packages/salesforce-adapter/src/transformers/xml_transformer.ts:191]: `replace` already replaces only the first match when it's passed as a string argument and not a regex

File: packages/salesforce-adapter/src/transformers/xml_transformer.ts [hunk 4]
```diff
@@ -173,6 +183,137 @@ export const toRetrieveRequest = (files: ReadonlyArray<FileProperties>): Retriev
...
+  '.svg': 'SVGContent',
+}
+
+const auraTypeToFileSuffix: Record<string, string> = {
+  Application: '.app',
+  Component: '.cmp',
+  Event: '.evt',
+  Interface: '.intf',
+  Tokens: '.tokens',
+}
```
[the remark this answers is not in the document]
Comment (omrilit) [reply packages/salesforce-adapter/src/transformers/xml_transformer.ts:231]: Agreed f2f to leave it there for now

File: packages/salesforce-adapter/src/transformers/xml_transformer.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/salesforce-adapter/src/transformers/xml_transformer.ts:191]: keep forgetting that, such a strange language

**Comment by eladmallel** [comment]
Thanks for sharing all this context @omrilit!

I've noticed a consistent structural decision in LCB I am not sure why it's like that.
The elements in question are: `targets`, `targetConfigs` and `lwcResources`.

For all of the above I expected them to be lists, however in your screenshot they all have a single field with the singular of the same name, and that field's value is the list. Why do we have this extra layer?

**Comment by omrilit** [comment]
@eladmallel this was the way it was about to look in case we use the unknown type.
Since we turned this option down and chose to define the type manually, it'll look as you would expect.

File: packages/salesforce-adapter/src/transformers/salesforce_types.ts [hunk 5]
```diff
@@ -14,144 +14,309 @@
...
 
-const MAP_AND_LOCATION_SETTINGS = {
-  enableAddressAutoComplete: BuiltinTypes.BOOLEAN,
-  enableMapsAndLocation: BuiltinTypes.BOOLEAN,
-}
+const targetConfigType = new ObjectType({
+  elemID: new ElemID(SALESFORCE, 'TargetConfig'),
+  fields: {
+    [`${XML_ATTRIBUTE_PREFIX}targets`]: {
+      type: BuiltinTypes.STRING, // todo retrieved as string delimited by ',' but is a list
```
Comment (omrilit) [inline packages/salesforce-adapter/src/transformers/salesforce_types.ts:118]: In the retrieved XML of LWC, the `targets` attribute is returned as a string delimited with ',' and not as a list (same as `datasource` attribute above).
@ori-moisis do you think that I should have a dedicated filter for transforming it to & from list post-fetch & pre-deploy respectively? anyway it will be done in separate PR if the answer is yes.

File: packages/salesforce-adapter/e2e_test/adapter.test.ts [hunk 6]
```diff
@@ -916,6 +916,85 @@ describe('Salesforce adapter E2E with real account', () => {
       } as MetadataInfo)
     }
 
+    const verifyAuraDefinitionBundleExists = async (): Promise<void> => {
```
Comment (ori-moisis) [inline packages/salesforce-adapter/e2e_test/adapter.test.ts:919]: a few things:
1. This file is becoming (already is actually) way too big to navigate effectively, how about extracting the very large instance definitions to not be in the same place as the tests?
Generally maybe it is worth splitting the whole setup from the tests given how many lines are involved in the setup?
2. With the new code it seems like we'd be calling `deploy` a lot of times (and if I had to guess, this isn't the first `deploy` call we have in the setup), we should have one deploy in the setup both for API limit and for the time it takes to run e2e


2 is more important, but since it may require a bit of refactoring anyway I figured it could be a good opportunity to cleanup the file a bit

File: packages/salesforce-adapter/e2e_test/adapter.test.ts [hunk 7]
```diff
@@ -1170,6 +1251,28 @@ describe('Salesforce adapter E2E with real account', () => {
...
+      const lwc = findElements(result, 'LightningComponentBundle',
+        'testLightningComponentBundle')[0] as InstanceElement
+      expect(lwc.value[constants.INSTANCE_FULL_NAME_FIELD])
+        .toEqual('testLightningComponentBundle')
+      const lwcResource = makeArray(lwc.value.lwcResources?.lwcResource)
+        .find(resource => resource.filePath === 'lwc/testLightningComponentBundle/testLightningComponentBundle.js')
+      expect(lwcResource).toBeDefined()
+      expect(isStaticFile(lwcResource.source)).toBe(true)
+      expect((lwcResource.source as StaticFile).content?.toString())
+        .toEqual("import { LightningElement } from 'lwc';\nexport default class BikeCard extends LightningElement {\n   name = 'Electra X4';\n   description = 'A sweet bike built for comfort.';\n   category = 'Mountain';\n   material = 'Steel';\n   price = '$2,700';\n   pictureUrl = 'https://s3-us-west-1.amazonaws.com/sfdc-demo/ebikes/electrax4.jpg';\n }")
```
Comment (ori-moisis) [inline packages/salesforce-adapter/e2e_test/adapter.test.ts:1274]: this large block of text should not be repeated IMO, should use the same definition that was used to create the instance on SFDC in the setup code

File: packages/salesforce-adapter/src/filters/missing_fields.json [hunk 8]
```diff
@@ -840,6 +840,34 @@
...
+        "type": "STRING",
+        "annotations": {
+          "_required": true,
+          "_values": [
+            "Application",
+            "Component",
+            "Event",
+            "Interface",
+            "Tokens"
+          ]
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/missing_fields.json:857]: not how the restriction annotation should be formatted

File: packages/salesforce-adapter/src/transformers/salesforce_types.ts [hunk 9]
```diff
@@ -14,144 +14,309 @@
...
-}
+const subTypesPath = [SALESFORCE, TYPES_PATH, SUBTYPES_PATH]
 
-const ACCOUNT_INTELLIGENCE_SETTINGS = {
-  enableAccountLogos: BuiltinTypes.BOOLEAN,
-  enableAutomatedAccountFields: BuiltinTypes.BOOLEAN,
-  enableNewsStories: BuiltinTypes.BOOLEAN,
-}
+const objectType = new ObjectType({
+  elemID: new ElemID(SALESFORCE, 'Object'),
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/transformers/salesforce_types.ts:26]: Names of subtypes in SFDC tend to have their parent as a prefix (like `LayoutSection`, `LayoutColumn`, etc...)
in this case, `salesforce.Object` is waaaay to generic IMO, it is a subtype, and should be named in a more specific way.
This applies to all the new subtypes you added here, not just the super generic sounding ones.

File: packages/salesforce-adapter/src/transformers/salesforce_types.ts [hunk 10]
```diff
@@ -14,144 +14,309 @@
...
+    },
+    [`${XML_ATTRIBUTE_PREFIX}role`]: {
+      type: BuiltinTypes.STRING,
+    },
+    [`${XML_ATTRIBUTE_PREFIX}type`]: {
+      type: BuiltinTypes.STRING,
+      annotations: {
+        [CORE_ANNOTATIONS.RESTRICTION]: createRestriction({
+          values: ['Boolean', 'Integer', 'String', 'Color', 'Date', 'DateTime'],
+          // eslint-disable-next-line @typescript-eslint/camelcase
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/transformers/salesforce_types.ts:74]: lint disable should not be needed here (this is assigning a value to a type that was already defined, I think the lint warning shows up only when you try to define something new with non camel case properties)

File: packages/salesforce-adapter/src/transformers/salesforce_types.ts (hunk 5 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/src/transformers/salesforce_types.ts:118]: I know it is a seemingly small effort, but it is code we have to maintain on a spec that isn't guaranteed, and the benefit of having it as a list of strings instead of a string with commas IMO is really really small.

Only reason I can think of that would justify the effort of making this transformation to a list is if we also make the list items references, but again, this feels like overkill to me

File: packages/salesforce-adapter/src/transformers/transformer.ts [hunk 11]
```diff
@@ -699,23 +695,8 @@ export class Types {
...
-      ),
-      path: [SALESFORCE, TYPES_PATH, ...isSubType ? [SUBTYPES_PATH] : [], typeName],
-    }))
-
   static getAllMissingTypes(): TypeElement[] {
-    return ([] as TypeElement[]).concat(
-      Types.generateMissingTypes(allMissingTypes),
-      Types.generateMissingTypes(allMissingSubTypes, true)
-    )
+    return ([] as TypeElement[]).concat(allMissingTypes, allMissingSubTypes)
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/transformers/transformer.ts:699]:
```suggestion
    return [...allMissingTypes, ...allMissingSubTypes]
```

File: packages/salesforce-adapter/src/transformers/transformer.ts [hunk 12]
```diff
@@ -943,8 +932,7 @@ export const transformPrimitive: TransformFunc = ({ value, path, field }) => {
   if (value === '') {
     return undefined
   }
-  // Salesforce returns nulls as objects like { $: { 'xsi:nil': 'true' } }
-  if ((_.isObject(value) && _.get(value, ['$', 'xsi:nil']) === 'true')) {
+  if (isXsiNil(value)) {
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/transformers/transformer.ts:935]: nits:
1. I think the function should be called `isNull` or `isSFDCNull` because that is the meaning of what it checks, the `xsi:nil` thing is an implementation detail
2. Maybe worth putting this in the same condition as `_.isNull` since it basically does the same thing?

File: packages/salesforce-adapter/src/transformers/xml_transformer.ts [hunk 13]
```diff
@@ -173,6 +183,137 @@ export const toRetrieveRequest = (files: ReadonlyArray<FileProperties>): Retriev
...
+  /**
+   * AuraDefinitionBundle has several base64Binary content fields. We should use its content fields
+   * suffix in order to set their content to the correct field
+   */
+  AuraDefinitionBundle: {
+    addContentFields: (fileNameToContent: Record<string, Buffer>, values: Values) => {
+      Object.entries(fileNameToContent)
+        .forEach(([contentFileName, content]) => {
+          const fieldName = Object.entries(auraFileSuffixToFieldName)
+            .find(([fileSuffix, _fieldName]) => contentFileName.endsWith(fileSuffix))?.[1]
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/salesforce-adapter/src/transformers/xml_transformer.ts:243]: resolved by not actually done, why?

File: packages/salesforce-adapter/src/transformers/xml_transformer.ts [hunk 14]
```diff
@@ -182,36 +323,52 @@ export const fromRetrieveResult = async (
...
+      if (!complexType) { // this is a single file
+        const zipFile = zip.file(`${PACKAGE}/${fileName}${withMetadataSuffix ? METADATA_XML_SUFFIX : ''}`)
+        return zipFile === null ? {} : { [zipFile.name]: await zipFile.async('nodebuffer') }
+      }
+      // bring all matching files from the fileName directory
+      const zipFiles = zip.file(new RegExp(`^${PACKAGE}/${fileName}/.*`))
+        .filter(zipFile => zipFile.name.endsWith(METADATA_XML_SUFFIX) === withMetadataSuffix)
+      return _.isEmpty(zipFiles)
+        ? {}
+        : _.fromPairs(await Promise.all(zipFiles.map(async zipFile => [zipFile.name, await zipFile.async('nodebuffer')])))
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/transformers/xml_transformer.ts:337]: relevant in a lot of places:
prefer built in language features over using lodash
```suggestion
        : Object.fromEntries(await Promise.all(zipFiles.map(async zipFile => [zipFile.name, await zipFile.async('nodebuffer')])))
```

File: packages/salesforce-adapter/src/transformers/xml_transformer.ts [hunk 15]
```diff
@@ -182,36 +323,52 @@ export const fromRetrieveResult = async (
...
-    const instanceFilename = typesWithMetaFile.has(file.type)
-      ? `${file.fileName}${METADATA_XML_SUFFIX}`
-      : file.fileName
-    const instanceData = await getFileData(instanceFilename)
-    if (instanceData === undefined) {
+    // extract metadata values
+    const fileNameToValuesBuffer = await extractFileNameToData(file.fileName,
+      typesWithMetaFile.has(file.type) || isComplexType(file.type), isComplexType(file.type))
+    if (Object.values(fileNameToValuesBuffer).length !== 1) {
+      log.warn(`Expected to retrieve only single values file for ${file.type}`)
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/transformers/xml_transformer.ts:344]: I am really confused about what is going on here.
It is really hard to try and follow the "common" flow when it is mixed in with the "complex" flow like this.
on the one hand I don't like that adding these "complex types" made the normal code flow so much more complicated and I believe a more clear separation would make both flows easier to read, on the other hand I am aware that I will probably re-write these parts of the code soon anyway so I am not sure if it would be wise to invest more time in them now...

I am really not sure what to do with this comment so I'll leave it up to you @omrilit - do you think it would be worth trying to make these different flows better separated or would you prefer that I try to do it (and might require some help understanding later on) when I refactor this file?

File: packages/salesforce-adapter/src/transformers/salesforce_types.ts (hunk 9 above)
Comment (omrilit) [reply packages/salesforce-adapter/src/transformers/salesforce_types.ts:26]: @ori-moisis Would you prefix it with the whole LightningComponentBundle? Only TargetConfigs? Both?
1. LightningComponentBundleTargetConfigObject
2. LightningComponentBundleObject
3. TargetConfigObject

File: packages/salesforce-adapter/src/transformers/salesforce_types.ts (hunk 10 above)
Comment (omrilit) [reply packages/salesforce-adapter/src/transformers/salesforce_types.ts:74]: It's still needed. You don't see it in other places since the lint disable is defined in the top of the file

File: packages/salesforce-adapter/src/transformers/xml_transformer.ts [hunk 16]
```diff
@@ -173,6 +183,137 @@ export const toRetrieveRequest = (files: ReadonlyArray<FileProperties>): Retriev
...
+      filepath: fileName.replace(PACKAGE, SALESFORCE),
+      content,
+    }))
+}
+
+type ComplexType = {
+  addContentFields(fileNameToContent: Record<string, Buffer>, values: Values): void
+  getMissingFields?(metadataFileName: string): Values
+  mapContentFields(instanceName: string, values: Values):
+    Record</* fieldName */string, Record</* fileName */string, /* content */string>>
```
[the remark this answers is not in the document]
Comment (omrilit) [reply packages/salesforce-adapter/src/transformers/xml_transformer.ts:200]: Done the map suggestion.
The content is string since it arrives from the instance's values, where it defined as string

File: packages/salesforce-adapter/src/transformers/xml_transformer.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (omrilit) [reply packages/salesforce-adapter/src/transformers/xml_transformer.ts:243]: I wrote it in the commit message when I reverted it.

> revert a PR fix of using path.ext since auraFileSuffixToFieldName has keys that are not extension

File: packages/salesforce-adapter/src/transformers/salesforce_types.ts (hunk 9 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/src/transformers/salesforce_types.ts:26]: either 1 or 2

File: packages/salesforce-adapter/src/transformers/xml_transformer.ts (hunk 15 above)
Comment (omrilit) [reply packages/salesforce-adapter/src/transformers/xml_transformer.ts:344]: I think that it'll be a waste of time investing on this one right now and once you'll start implementing the new flow I can take the separation of this logic or help you understanding it

File: packages/salesforce-adapter/e2e_test/adapter.test.ts (hunk 6 above)
Comment (omrilit) [reply packages/salesforce-adapter/e2e_test/adapter.test.ts:919]: @ori-moisis 
1. Done
2. There was only a single call to deploy and I have added Two more. IMO this refactor is indeed important but should be done after the adapter will support a deploy for multiple instances at the same time.

File: packages/salesforce-adapter/src/transformers/salesforce_types.ts (hunk 5 above)
Comment (omrilit) [reply packages/salesforce-adapter/src/transformers/salesforce_types.ts:118]: Created a related ticket at SALTO-861

File: packages/salesforce-adapter/e2e_test/adapter.test.ts (hunk 6 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/e2e_test/adapter.test.ts:919]: this has nothing do with the the adapter though, the setup code is using `SalesforceClient`

Comment (omrilit) [reply packages/salesforce-adapter/e2e_test/adapter.test.ts:919]: It uses `toMetadataPackageZip` in order to prepare the zip, but today it takes only a single instance

Comment (ori-moisis) [reply packages/salesforce-adapter/e2e_test/adapter.test.ts]: got it, I'll take care of it then

File: packages/salesforce-adapter/e2e_test/adapter.test.ts [hunk 17]
```diff
@@ -3836,14 +3088,127 @@ describe('Salesforce adapter E2E with real account', () => {
...
           })
         })
+
+        describe('AuraDefinitionBundle manipulation', () => {
+          let auraInstance: InstanceElement
+
+          beforeAll(async () => {
+            auraInstance = await createInstance(client, {
+              [constants.INSTANCE_FULL_NAME_FIELD]: 'MyAuraDefinitionBundle',
+              SVGContent: '<?xml version="1.0" encoding="UTF-8" standalone="no"?>\n<svg width="120px" height="120px" viewBox="0 0 120 120" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">\n\t<g stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">\n\t\t<path d="M120,108 C120,114.6 114.6,120 108,120 L12,120 C5.4,120 0,114.6 0,108 L0,12 C0,5.4 5.4,0 12,0 L108,0 C114.6,0 120,5.4 120,12 L120,108 L120,108 Z" id="Shape" fill="#2A739E"/>\n\t\t<path d="M77.7383308,20 L61.1640113,20 L44.7300055,63.2000173 L56.0543288,63.2000173 L40,99.623291 L72.7458388,54.5871812 L60.907727,54.5871812 L77.7383308,20 Z" id="Path-1" fill="#FFFFFF"/>\n\t</g>\n</svg>',
```
Comment (ori-moisis) [inline packages/salesforce-adapter/e2e_test/adapter.test.ts:3106]: nit: can probably reuse most of the bundle definition from `setup.ts` instead of re-writing it here, even if we do need to create a new instance, if it has the same values and just the name is different, it would be easier and more readable IMO to copy the original instance definition and just change the name(s)

File: packages/salesforce-adapter/src/transformers/transformer.ts [hunk 18]
```diff
@@ -935,16 +916,20 @@ const convertXsdTypeFuncMap: Record<string, ConvertXsdTypeFunc> = {
   'xsd:long': Number,
 }
 
+// Salesforce returns nulls in metadata API as objects like { $: { 'xsi:nil': 'true' } }
+// and in retrieve API like <activateRSS xsi:nil="true"/>
+// which is transformed to { `${XML_ATTRIBUTE_PREFIX}xsi:nil`): 'true' }
+const isNull = (value: Value): boolean =>
+  _.isNull(value) || (_.isObject(value)
+    && (_.get(value, ['$', 'xsi:nil']) === 'true'
+      || _.get(value, `${XML_ATTRIBUTE_PREFIX}xsi:nil`) === 'true'))
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/transformers/transformer.ts:925]: something I just realized and now I wonder if it would be better.
I am pretty sure the `$` thing is the way `jsforce` treats xml attributes. maybe it would be better to do the same (nest all attributes under a common key instead of adding a prefix to each one).
Ideally I would even say we should align our handling with jsforce so we won't have to treat things differently because our xml parser is different than jsforce's xml parser.
wdyt?

Comment (omrilit) [reply packages/salesforce-adapter/src/transformers/transformer.ts:925]: @ori-moisis  I can define all the attributes under the `$` field and than the `xsi:nil` thingy will be caught as it does from `jsforce`.
The problem with this approach is that I need to define the type of the `targetConfigs` inner fields under  the `$` fields and our parser doesn't handle `$`.
So... as I see it, we can either:
1. leave it as is
2. change the parser to support `$`and modify the temporary inner types of `targetConfigs` to have the `$` field. Once we have the filter that modifies the attributes of `targetConfig`, there should not be any fields with `$` and if they will appear we will anyway filter them out in `convert_types` filter.

IMO it doesn't worth the effort. WDYT?

File: packages/salesforce-adapter/e2e_test/adapter.test.ts (hunk 17 above)
Comment (omrilit) [reply packages/salesforce-adapter/e2e_test/adapter.test.ts]: Done

File: packages/salesforce-adapter/src/transformers/transformer.ts (hunk 18 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/src/transformers/transformer.ts:925]: 3. use a different name that the parser does support ("attributes"?)

anyway, up to you, not critical right now IMO
